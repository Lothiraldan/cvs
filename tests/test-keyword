#!/bin/sh

cat <<EOF >> $HGRCPATH
[extensions]
keyword =
mq =
notify =
record =
transplant =
[ui]
interactive = true
EOF

# demo before [keyword] files are set up
# would succeed without uisetup otherwise
echo % hg kwdemo
hg --quiet kwdemo \
| sed -e 's![^ ][^ ]*demo.txt,v!/TMP/demo.txt,v!' \
 -e 's/,v [a-z0-9][a-z0-9]* /,v xxxxxxxxxxxx /' \
 -e '/[$]Revision/ s/: [a-z0-9][a-z0-9]* /: xxxxxxxxxxxx /' \
 -e 's! 20[0-9][0-9]/[01][0-9]/[0-3][0-9] [0-2][0-9]:[0-6][0-9]:[0-6][0-9]! 2000/00/00 00:00:00!'

hg --quiet kwdemo "Branch = {branches}"

cat <<EOF >> $HGRCPATH
[keyword]
** =
b = ignore
[hooks]
commit=
commit.test=cp a hooktest
EOF

hg init Test-bndl
cd Test-bndl

echo % kwshrink should exit silently in empty/invalid repo
hg kwshrink

# Symlinks cannot be created on Windows. The bundle was made with:
#
# hg init t
# cd t
# echo a > a
# ln -s a sym
# hg add sym
# hg ci -m addsym -u mercurial
# hg bundle --base null ../test-keyword.hg
#
hg pull -u "$TESTDIR/test-keyword.hg" \
    | sed 's/pulling from.*test-keyword.hg/pulling from test-keyword.hg/'

echo 'expand $Id$' > a
echo 'do not process $Id:' >> a
echo 'xxx $' >> a
echo 'ignore $Id$' > b
echo % cat
cat a b

echo % no kwfiles
hg kwfiles
echo % untracked candidates
hg -v kwfiles --unknown

echo % addremove
hg addremove
echo % status
hg status

echo % default keyword expansion including commit hook
echo % interrupted commit should not change state or run commit hook
hg --debug commit
echo % status
hg status

echo % commit
hg --debug commit -mabsym -u 'User Name <user@example.com>'
echo % status
hg status
echo % identify
hg debugrebuildstate
hg --quiet identify
echo % cat
cat a b
echo % hg cat
hg cat sym a b

echo
echo % diff a hooktest
diff a hooktest

echo % removing commit hook from config
sed -e '/\[hooks\]/,$ d' "$HGRCPATH" > $HGRCPATH.nohook
mv "$HGRCPATH".nohook "$HGRCPATH"
rm hooktest

echo % bundle
hg bundle --base null ../kw.hg

cd ..
hg init Test
cd Test

echo % notify on pull to check whether keywords stay as is in email
echo % ie. if patch.diff wrapper acts as it should

cat <<EOF >> $HGRCPATH
[hooks]
incoming.notify = python:hgext.notify.hook
[notify]
sources = pull
diffstat = False
[reposubs]
* = Test
EOF

echo % pull from bundle
hg pull -u ../kw.hg 2>&1 | sed -e '/^Content-Type:/,/^diffs (/ d'

echo % remove notify config
sed -e '/\[hooks\]/,$ d' "$HGRCPATH" > $HGRCPATH.nonotify
mv "$HGRCPATH".nonotify "$HGRCPATH"

echo % touch
touch a b
echo % status
hg status

rm sym a b
echo % update
hg update -C
echo % cat
cat a b

echo % check whether expansion is filewise
echo '$Id$' > c
echo 'tests for different changenodes' >> c
echo % commit c
hg commit -A -mcndiff -d '1 0' -u 'User Name <user@example.com>'
echo % force expansion
hg -v kwexpand
echo % compare changenodes in a c
cat a c

echo % record
cp "$HGRCPATH" $HGRCPATH.bak
sed -e '1 a \foo' a > a.tmp
mv a.tmp a
echo bar >> a
hg record -d '1 10' -m rectest<<EOF
y
y
n
EOF
echo
hg identify
hg status
echo % cat modified file
cat a
hg diff | grep -v 'b/a'
hg rollback
hg update -C

echo % init --mq
hg init --mq
echo % qimport
hg qimport -r tip -n mqtest.diff
echo % commit --mq
hg commit --mq -m mqtest
echo % keywords should not be expanded in patch
cat .hg/patches/mqtest.diff
echo % qpop
hg qpop
echo % qgoto - should imply qpush
hg qgoto mqtest.diff
echo % cat
cat c
echo % hg cat
hg cat c
echo % keyword should not be expanded in filelog
hg --config 'extensions.keyword=!' cat c
echo % qpop and move on
hg qpop

echo % copy
hg cp a c

echo % kwfiles added
hg kwfiles

echo % commit
hg --debug commit -ma2c -d '1 0' -u 'User Name <user@example.com>'
echo % cat a c
cat a c
echo % touch copied c
touch c
echo % status
hg status

echo % kwfiles
hg kwfiles
echo % ignored files
hg -v kwfiles --ignore
echo % all files
hg kwfiles --all

echo % diff --rev
hg diff --rev 1 | grep -v 'b/c'

echo % rollback
hg rollback
echo % status
hg status
echo % update -C
hg update --clean

echo % custom keyword expansion
echo % try with kwdemo
hg --quiet kwdemo "Xinfo = {author}: {desc}"

cat <<EOF >>$HGRCPATH
[keywordmaps]
Id = {file} {node|short} {date|rfc822date} {author|user}
Xinfo = {author}: {desc}
EOF

echo % cat
cat a b
echo % hg cat
hg cat sym a b

echo
echo '$Xinfo$' >> a
cat <<EOF >> log
firstline
secondline
EOF

echo % interrupted commit should not change state
hg commit
echo % status
hg status

echo % commit
hg --debug commit -l log -d '2 0' -u 'User Name <user@example.com>'
rm log
echo % status
hg status
echo % verify
hg verify

echo % cat
cat a b
echo % hg cat
hg cat sym a b
echo
echo % annotate
hg annotate a

echo % remove
hg debugrebuildstate
hg remove a
hg --debug commit -m rma
echo % status
hg status
echo % rollback
hg rollback
echo % status
hg status
echo % revert a
hg revert --no-backup --rev tip a
echo % cat a
cat a

echo % clone to test incoming
cd ..
hg clone -r1 Test Test-a
cd Test-a
cat <<EOF >> .hg/hgrc
[paths]
default = ../Test
EOF
echo % incoming
# remove path to temp dir
hg incoming | sed -e 's/^\(comparing with \).*\(test-keyword.*\)/\1\2/'

sed -e 's/Id.*/& rejecttest/' a > a.new
mv a.new a
echo % commit rejecttest
hg --debug commit -m'rejects?' -d '3 0' -u 'User Name <user@example.com>'
echo % export
hg export -o ../rejecttest.diff tip

cd ../Test
echo % import
hg import ../rejecttest.diff
echo % cat
cat a b
echo
echo % rollback
hg rollback
echo % clean update
hg update --clean

echo % kwexpand/kwshrink on selected files
mkdir x
echo % copy a x/a
hg copy a x/a
echo % kwexpand a
hg --verbose kwexpand a
echo % kwexpand x/a should abort
hg --verbose kwexpand x/a
cd x
hg --debug commit -m xa -d '3 0' -u 'User Name <user@example.com>'
echo % cat a
cat a
echo % kwshrink a inside directory x
hg --verbose kwshrink a
echo % cat a
cat a
cd ..

echo % kwexpand nonexistent
hg kwexpand nonexistent 2>&1 | sed 's/nonexistent:.*/nonexistent:/'

echo % hg serve
hg serve -p $HGPORT -d --pid-file=hg.pid -A access.log -E errors.log
cat hg.pid >> $DAEMON_PIDS
echo % expansion
echo % hgweb file
("$TESTDIR/get-with-headers.py" localhost:$HGPORT '/file/tip/a/?style=raw')
echo % no expansion
echo % hgweb annotate
("$TESTDIR/get-with-headers.py" localhost:$HGPORT '/annotate/tip/a/?style=raw')
echo % hgweb changeset
("$TESTDIR/get-with-headers.py" localhost:$HGPORT '/rev/tip/?style=raw')
echo % hgweb filediff
("$TESTDIR/get-with-headers.py" localhost:$HGPORT '/diff/bb948857c743/a?style=raw')
echo % errors encountered
cat errors.log

echo % merge/resolve
echo '$Id$' > m
hg add m
hg commit -m 4kw 
echo foo >> m
hg commit -m 5foo
echo % simplemerge
hg update 4
echo foo >> m
hg commit -m 6foo
hg merge
hg commit -m simplemerge
cat m
echo % conflict
hg update 4
echo bar >> m
hg commit -m 8bar
hg merge
echo % keyword stays outside conflict zone
cat m
echo % resolve to local
HGMERGE=internal:local hg resolve -a
hg commit -m localresolve
cat m

echo % test restricted mode with transplant -b
hg update 6
hg branch foo
mv a a.bak
echo foobranch > a
cat a.bak >> a
rm a.bak
hg commit -m 9foobranch
hg update default
hg -y transplant -b foo tip
echo % no expansion in changeset
hg tip -p
echo % expansion in file
head -n 2 a
hg -q rollback
hg -q update -C

echo % switch off expansion
echo % kwshrink with unknown file u
cp a u
hg --verbose kwshrink
echo % cat
cat a b
echo % hg cat
hg cat sym a b
echo
rm "$HGRCPATH"
echo % cat
cat a b
echo % hg cat
hg cat sym a b
echo
