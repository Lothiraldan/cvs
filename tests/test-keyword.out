% hg kwdemo
[extensions]
keyword = 
[keyword]
demo.txt = 
[keywordmaps]
Author = {author|user}
Date = {date|utcdate}
Header = {root}/{file},v {node|short} {date|utcdate} {author|user}
Id = {file|basename},v {node|short} {date|utcdate} {author|user}
RCSFile = {file|basename},v
RCSfile = {file|basename},v
Revision = {node|short}
Source = {root}/{file},v
$Author: test $
$Date: 2000/00/00 00:00:00 $
$Header: /TMP/demo.txt,v xxxxxxxxxxxx 2000/00/00 00:00:00 test $
$Id: demo.txt,v xxxxxxxxxxxx 2000/00/00 00:00:00 test $
$RCSFile: demo.txt,v $
$RCSfile: demo.txt,v $
$Revision: xxxxxxxxxxxx $
$Source: /TMP/demo.txt,v $
[extensions]
keyword = 
[keyword]
demo.txt = 
[keywordmaps]
Branch = {branches}
$Branch: demobranch $
% kwshrink should exit silently in empty/invalid repo
pulling from test-keyword.hg
requesting all changes
adding changesets
adding manifests
adding file changes
added 1 changesets with 1 changes to 1 files
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
% cat
expand $Id$
do not process $Id:
xxx $
ignore $Id$
% no kwfiles
% untracked candidates
k a
% addremove
adding a
adding b
% status
A a
A b
% default keyword expansion including commit hook
% interrupted commit should not change state or run commit hook
abort: empty commit message
% status
A a
A b
% commit
a
b
overwriting a expanding keywords
running hook commit.test: cp a hooktest
committed changeset 1:ef63ca68695bc9495032c6fda1350c71e6d256e9
% status
? hooktest
% identify
ef63ca68695b
% cat
expand $Id: a,v ef63ca68695b 1970/01/01 00:00:00 user $
do not process $Id:
xxx $
ignore $Id$
% hg cat
expand $Id: a,v ef63ca68695b 1970/01/01 00:00:00 user $
do not process $Id:
xxx $
ignore $Id$
a
% diff a hooktest
% removing commit hook from config
% bundle
2 changesets found
% notify on pull to check whether keywords stay as is in email
% ie. if patch.diff wrapper acts as it should
% pull from bundle
pulling from ../kw.hg
requesting all changes
adding changesets
adding manifests
adding file changes
added 2 changesets with 3 changes to 3 files

diff -r 000000000000 -r a2392c293916 sym
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/sym	Sat Feb 09 20:25:47 2008 +0100
@@ -0,0 +1,1 @@
+a
\ No newline at end of file

diff -r a2392c293916 -r ef63ca68695b a
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/a	Thu Jan 01 00:00:00 1970 +0000
@@ -0,0 +1,3 @@
+expand $Id$
+do not process $Id:
+xxx $
diff -r a2392c293916 -r ef63ca68695b b
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/b	Thu Jan 01 00:00:00 1970 +0000
@@ -0,0 +1,1 @@
+ignore $Id$
3 files updated, 0 files merged, 0 files removed, 0 files unresolved
% remove notify config
% touch
% status
% update
3 files updated, 0 files merged, 0 files removed, 0 files unresolved
% cat
expand $Id: a,v ef63ca68695b 1970/01/01 00:00:00 user $
do not process $Id:
xxx $
ignore $Id$
% check whether expansion is filewise
% commit c
adding c
% force expansion
overwriting a expanding keywords
overwriting c expanding keywords
% compare changenodes in a c
expand $Id: a,v ef63ca68695b 1970/01/01 00:00:00 user $
do not process $Id:
xxx $
$Id: c,v 40a904bbbe4c 1970/01/01 00:00:01 user $
tests for different changenodes
% qinit -c
% qimport
% qcommit
% keywords should not be expanded in patch
# HG changeset patch
# User User Name <user@example.com>
# Date 1 0
# Node ID 40a904bbbe4cd4ab0a1f28411e35db26341a40ad
# Parent  ef63ca68695bc9495032c6fda1350c71e6d256e9
cndiff

diff -r ef63ca68695b -r 40a904bbbe4c c
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/c	Thu Jan 01 00:00:01 1970 +0000
@@ -0,0 +1,2 @@
+$Id$
+tests for different changenodes
% qpop
popping mqtest.diff
patch queue now empty
% qgoto - should imply qpush
applying mqtest.diff
now at: mqtest.diff
% cat
$Id: c,v 40a904bbbe4c 1970/01/01 00:00:01 user $
tests for different changenodes
% qpop and move on
popping mqtest.diff
patch queue now empty
% copy
% kwfiles added
a
c
% commit
c
 c: copy a:0045e12f6c5791aac80ca6cbfd97709a88307292
overwriting c expanding keywords
committed changeset 2:25736cf2f5cbe41f6be4e6784ef6ecf9f3bbcc7d
% cat a c
expand $Id: a,v ef63ca68695b 1970/01/01 00:00:00 user $
do not process $Id:
xxx $
expand $Id: c,v 25736cf2f5cb 1970/01/01 00:00:01 user $
do not process $Id:
xxx $
% touch copied c
% status
% kwfiles
a
c
% ignored files
I b
I sym
% all files
K a
K c
I b
I sym
% diff --rev
diff -r ef63ca68695b c
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
@@ -0,0 +1,3 @@
+expand $Id$
+do not process $Id:
+xxx $
% rollback
rolling back last transaction
% status
A c
% update -C
0 files updated, 0 files merged, 0 files removed, 0 files unresolved
% custom keyword expansion
% try with kwdemo
[extensions]
keyword = 
[keyword]
* = 
b = ignore
demo.txt = 
[keywordmaps]
Xinfo = {author}: {desc}
$Xinfo: test: hg keyword config and expansion example $
% cat
expand $Id: a,v ef63ca68695b 1970/01/01 00:00:00 user $
do not process $Id:
xxx $
ignore $Id$
% hg cat
expand $Id: a ef63ca68695b Thu, 01 Jan 1970 00:00:00 +0000 user $
do not process $Id:
xxx $
ignore $Id$
a
% interrupted commit should not change state
abort: empty commit message
% status
M a
? c
? log
% commit
a
overwriting a expanding keywords
committed changeset 2:bb948857c743469b22bbf51f7ec8112279ca5d83
% status
? c
% verify
checking changesets
checking manifests
crosschecking files in changesets and manifests
checking files
3 files, 3 changesets, 4 total revisions
% cat
expand $Id: a bb948857c743 Thu, 01 Jan 1970 00:00:02 +0000 user $
do not process $Id:
xxx $
$Xinfo: User Name <user@example.com>: firstline $
ignore $Id$
% hg cat
expand $Id: a bb948857c743 Thu, 01 Jan 1970 00:00:02 +0000 user $
do not process $Id:
xxx $
$Xinfo: User Name <user@example.com>: firstline $
ignore $Id$
a
% annotate
1: expand $Id$
1: do not process $Id:
1: xxx $
2: $Xinfo$
% remove
committed changeset 3:d14c712653769de926994cf7fbb06c8fbd68f012
% status
? c
% rollback
rolling back last transaction
% status
R a
? c
% revert a
% cat a
expand $Id: a bb948857c743 Thu, 01 Jan 1970 00:00:02 +0000 user $
do not process $Id:
xxx $
$Xinfo: User Name <user@example.com>: firstline $
% clone to test incoming
requesting all changes
adding changesets
adding manifests
adding file changes
added 2 changesets with 3 changes to 3 files
updating to branch default
3 files updated, 0 files merged, 0 files removed, 0 files unresolved
% incoming
comparing with test-keyword/Test
searching for changes
changeset:   2:bb948857c743
tag:         tip
user:        User Name <user@example.com>
date:        Thu Jan 01 00:00:02 1970 +0000
summary:     firstline

% commit rejecttest
a
overwriting a expanding keywords
committed changeset 2:85e279d709ffc28c9fdd1b868570985fc3d87082
% export
% import
applying ../rejecttest.diff
% cat
expand $Id: a 4e0994474d25 Thu, 01 Jan 1970 00:00:03 +0000 user $ rejecttest
do not process $Id: rejecttest
xxx $
$Xinfo: User Name <user@example.com>: rejects? $
ignore $Id$

% rollback
rolling back last transaction
% clean update
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
% kwexpand/kwshrink on selected files
% copy a x/a
% kwexpand a
overwriting a expanding keywords
% kwexpand x/a should abort
abort: outstanding uncommitted changes
x/a
 x/a: copy a:779c764182ce5d43e2b1eb66ce06d7b47bfe342e
overwriting x/a expanding keywords
committed changeset 3:b4560182a3f9a358179fd2d835c15e9da379c1e4
% cat a
expand $Id: x/a b4560182a3f9 Thu, 01 Jan 1970 00:00:03 +0000 user $
do not process $Id:
xxx $
$Xinfo: User Name <user@example.com>: xa $
% kwshrink a inside directory x
overwriting x/a shrinking keywords
% cat a
expand $Id$
do not process $Id:
xxx $
$Xinfo$
% kwexpand nonexistent
nonexistent:
% hg serve
% expansion
% hgweb file
200 Script output follows

expand $Id: a bb948857c743 Thu, 01 Jan 1970 00:00:02 +0000 user $
do not process $Id:
xxx $
$Xinfo: User Name <user@example.com>: firstline $
% no expansion
% hgweb annotate
200 Script output follows


user@1: expand $Id$
user@1: do not process $Id:
user@1: xxx $
user@2: $Xinfo$




% hgweb changeset
200 Script output follows


# HG changeset patch
# User User Name <user@example.com>
# Date 3 0
# Node ID b4560182a3f9a358179fd2d835c15e9da379c1e4
# Parent  bb948857c743469b22bbf51f7ec8112279ca5d83
xa

diff -r bb948857c743 -r b4560182a3f9 x/a
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/x/a	Thu Jan 01 00:00:03 1970 +0000
@@ -0,0 +1,4 @@
+expand $Id$
+do not process $Id:
+xxx $
+$Xinfo$

% hgweb filediff
200 Script output follows


diff -r ef63ca68695b -r bb948857c743 a
--- a/a	Thu Jan 01 00:00:00 1970 +0000
+++ b/a	Thu Jan 01 00:00:02 1970 +0000
@@ -1,3 +1,4 @@
 expand $Id$
 do not process $Id:
 xxx $
+$Xinfo$




% errors encountered
% merge/resolve
% simplemerge
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
created new head
0 files updated, 0 files merged, 0 files removed, 0 files unresolved
(branch merge, don't forget to commit)
$Id: m 27d48ee14f67 Thu, 01 Jan 1970 00:00:00 +0000 test $
foo
% conflict
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
created new head
merging m
warning: conflicts during merge.
merging m failed!
0 files updated, 0 files merged, 0 files removed, 1 files unresolved
use 'hg resolve' to retry unresolved file merges or 'hg update -C' to abandon
% keyword stays outside conflict zone
$Id$
<<<<<<< local
bar
=======
foo
>>>>>>> other
% resolve to local
$Id: m 41efa6d38e9b Thu, 01 Jan 1970 00:00:00 +0000 test $
bar
% test restricted mode with transplant -b
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
marked working directory as branch foo
created new head
2 files updated, 0 files merged, 0 files removed, 0 files unresolved
applying 4aa30d025d50
4aa30d025d50 transplanted to 5a4da427c162
% no expansion in changeset
changeset:   11:5a4da427c162
tag:         tip
parent:      9:41efa6d38e9b
user:        test
date:        Thu Jan 01 00:00:00 1970 +0000
summary:     9foobranch

diff -r 41efa6d38e9b -r 5a4da427c162 a
--- a/a	Thu Jan 01 00:00:00 1970 +0000
+++ b/a	Thu Jan 01 00:00:00 1970 +0000
@@ -1,3 +1,4 @@
+foobranch
 expand $Id$
 do not process $Id:
 xxx $

% expansion in file
foobranch
expand $Id: a 5a4da427c162 Thu, 01 Jan 1970 00:00:00 +0000 test $
% switch off expansion
% kwshrink with unknown file u
overwriting a shrinking keywords
overwriting m shrinking keywords
overwriting x/a shrinking keywords
% cat
expand $Id$
do not process $Id:
xxx $
$Xinfo$
ignore $Id$
% hg cat
expand $Id: a bb948857c743 Thu, 01 Jan 1970 00:00:02 +0000 user $
do not process $Id:
xxx $
$Xinfo: User Name <user@example.com>: firstline $
ignore $Id$
a
% cat
expand $Id$
do not process $Id:
xxx $
$Xinfo$
ignore $Id$
% hg cat
expand $Id$
do not process $Id:
xxx $
$Xinfo$
ignore $Id$
a
